tests/unit/test_hash.cpp

```diff
@@ -0,0 +1,143 @@
+const auto sha256_foobar = "c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2";
+
+TEST_F(HashTest, sha256EmptyString)
+{
+    std::string value;
+    Hash::sha256().digest(value);
+    EXPECT_THAT(value, Eq(sha256_emptyString));
+}
+
+TEST_F(HashTest, sha512EmptyString)
+{
+    std::string value;
+    Hash::sha512().digest(value);
+    EXPECT_THAT(value, Eq(sha512_emptyString));
+}
+
+TEST_F(HashTest, sha256SingleUpdateString)
+{
+    std::string value;
+    Hash::sha256().update("foo").digest(value);
+    EXPECT_THAT(value, Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256MultipleUpdatesString)
+{
+    std::string value;
+    Hash::sha256().update("foo").update("bar").digest(value);
+    EXPECT_THAT(value, Eq(sha256_foobar));
+}
+
+TEST_F(HashTest, sha256SingleUpdateBinaryArray)
+{
+    std::string value;
+    Hash::sha256().update(reinterpret_cast<const uint8_t*>(&"foo"[0]), 3).digest(value);
+    EXPECT_THAT(value, Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256SingleUpdateVector)
+{
+    std::string chunk = "foo";
+    std::vector<uint8_t> chunk2conv(chunk.begin(), chunk.end());
+    std::string value;
+    Hash::sha256().update(chunk2conv).digest(value);
+    EXPECT_THAT(value, Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256getDigestAsBinaryArray)
+{
+    uint8_t value[Hash::lengthBytes[openssl::DigestTypes::SHA256]];
+    unsigned int length;
+    Hash::sha256().update("foo").digest(value, length);
+    EXPECT_THAT(utility::toHex(value, length), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256getDigestAsVector)
+{
+    std::vector<uint8_t> value;
+    Hash::sha256().update("foo").digest(value);
+    EXPECT_THAT(utility::toHex(value), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionStringVersionReturnsHex)
+{
+    std::string digest;
+    sha256("foo", digest);
+    EXPECT_THAT(digest, Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionVectorVersion)
+{
+    std::string message = "foo";
+    std::vector<uint8_t> messageConv(message.begin(), message.end());
+    std::vector<uint8_t> digest;
+    sha256(messageConv, digest);
+    EXPECT_THAT(utility::toHex(digest), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionBinaryVersion)
+{
+    std::string message = "foo";
+    uint8_t digest[Hash::lengthBytes[openssl::DigestTypes::SHA256]];
+    unsigned int length;
+    sha256(reinterpret_cast<const uint8_t*>(message.c_str()), message.length(), digest, length);
+    EXPECT_THAT(utility::toHex(digest, length), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, returnsDigestLength) {
+    EXPECT_THAT(Hash::sha256().getDigestLengthBytes(), Eq(32));
+    EXPECT_THAT(Hash::sha512().getDigestLengthBytes(), Eq(64));
+}
+
+TEST_F(HashTest, throwsIfCalledUpdateAfterDigest)
+{
+    std::string value;
+    EXPECT_THROW({
+        Hash sha256 = Hash::sha256();
+        sha256.update("foo");
+        sha256.digest(value);
+        sha256.update("bar");
+    }, std::runtime_error);
+}
```

Comment:
Tests for toHex are missing. Moreover tests for the binary output of the hash functions are missing.
Yes, they are tested implicitly by the other tests but if such a test fails it gives you a wrong indication where the problem is and kind of negates what you want to achieve with testing (avoid lengthy debugging)

Reply:
added tests for toHex